mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/query/parsing/nativequery/impl/NativeQueryParser.java

```diff
@@ -66,7 +68,11 @@ public NativeQueryParser() {
 	}
 
 	public Rule Query() {
-		return Sequence( FirstOf( ParsedQuery(), CriteriaOnlyFindQuery() ), EOI, push( builder ) );
+		return Sequence( Optional( QueryCliOrSimple() ), push( builder ) );
```

Comment:
I'm a bit puzzled by this change, especially the `Optional`? It is to support returning an error for empty queries?

Reply:
Suppose the `Optional` is omitted and then we run the parser and `QueryCliOrSimple` fails. The `push( builder )` does not get executed and the value stack (and resultValue) is empty and there is no builder. 

If we swaped `push( builder )` and `QueryCliOrSimple` the value stack will be empty also because according to the [documentation](https://github.com/sirthias/parboiled/wiki/The-Value-Stack#rule-mismatches):

> If a rule did not match for whatever reason the value stack is reset to the exact state it had before the execution of the failing rule was started.

Reply:
Okay, I think we probably won't have an issue once the `push()` is at the beginning of everything (we will need that to make the parser stateless).